rclpy/rclpy/executors.py

```diff
@@ -305,93 +291,73 @@ def wait_for_ready_callbacks(self, timeout_sec=None, nodes=None):
-
-                entities = {
-                    'subscription': (subscriptions, 'subscription_handle'),
-                    'guard_condition': (guards, 'guard_handle'),
-                    'client': (clients, 'client_handle'),
-                    'service': (services, 'service_handle'),
-                    'timer': (timers, 'timer_handle'),
-                }
-                for entity, (handles, handle_name) in entities.items():
-                    _rclpy.rclpy_wait_set_clear_entities(entity, wait_set)
-                    for h in handles:
-                        _rclpy.rclpy_wait_set_add_entity(
-                            entity, wait_set, h.__getattribute__(handle_name)
-                        )
-                _rclpy.rclpy_wait_set_add_entity('guard_condition', wait_set, sigint_gc)
-                _rclpy.rclpy_wait_set_add_entity(
-                    'guard_condition', wait_set, self._guard_condition)
+                wait_set.add_subscriptions(subscriptions)
+                wait_set.add_clients(clients)
+                wait_set.add_services(services)
+                wait_set.add_timers(timers)
+                wait_set.add_guard_conditions(guards)
+                wait_set.add_guard_condition(sigint_gc, sigint_gc_handle)
+                wait_set.add_guard_condition(self._guard_condition, self._guard_condition_handle)
 
                 # Wait for something to become ready
-                _rclpy.rclpy_wait(wait_set, timeout_nsec)
-
-                # get ready entities
-                subs_ready = _rclpy.rclpy_get_ready_entities('subscription', wait_set)
-                guards_ready = _rclpy.rclpy_get_ready_entities('guard_condition', wait_set)
-                timers_ready = _rclpy.rclpy_get_ready_entities('timer', wait_set)
-                clients_ready = _rclpy.rclpy_get_ready_entities('client', wait_set)
-                services_ready = _rclpy.rclpy_get_ready_entities('service', wait_set)
+                wait_set.wait(timeout_nsec)
+
+                # Check sigint guard condition
+                if wait_set.is_ready(sigint_gc_handle):
+                    raise KeyboardInterrupt
+                _rclpy.rclpy_destroy_entity(sigint_gc)
+
+                # Mark all guards as triggered before yielding since they're auto-taken
+                for gc in [g for g in guards if wait_set.is_ready(g.guard_pointer)]:
+                    gc._executor_triggered = True
+
+                # Process ready entities one node at a time
+                for node in nodes:
+                    for tmr in [t for t in timers if wait_set.is_ready(t.timer_pointer)]:
+                        # Check that a timer is ready to workaround rcl issue with cancelled timers
+                        if _rclpy.rclpy_is_timer_ready(tmr.timer_handle):
+                            if tmr == timeout_timer:
+                                continue
+                            elif tmr.callback_group.can_execute(tmr):
```

Comment:
Nitpick: this doesn't need to be an `elif` since the if block calls `continue`. `if` would be sufficient.

Reply:
If it is alright with you I would like to leave this for now since it gets fixed in the next PR #135 

https://github.com/ros2/rclpy/pull/135/files#diff-521ddf9ec814122daa49f03bb2974340R245